rust/lance-index/src/vector/pq/distance.rs

```diff
@@ -93,6 +95,33 @@ pub(super) fn compute_l2_distance(
     distances
 }
 
+#[inline]
+pub(super) fn compute_l2_distance_4bit(
+    distance_table: &[f32],
+    num_sub_vectors: usize,
+    code: &[u8],
+) -> Vec<f32> {
+    let num_vectors = code.len() * 2 / num_sub_vectors;
+    let mut distances = vec![0.0_f32; num_vectors];
+    let num_centroids = 2_usize.pow(4);
+    for (sub_vec_idx, vec_indices) in code.chunks_exact(num_vectors).enumerate() {
+        let dist_table = &distance_table[sub_vec_idx * 2 * num_centroids..];
+        debug_assert_eq!(vec_indices.len(), distances.len());
+        vec_indices
+            .iter()
+            .zip(distances.iter_mut())
+            .for_each(|(&centroid_idx, sum)| {
+                // for 4bit PQ, `centroid_idx` is 2 index, each index is 4bit.
```

Comment:
How do we make sure the registers are not swap out in such case, giving we have very high PQ, i.e., PQ=96 for dim=768. Do we have enough registers?

Reply:
let's keep track of this as a potential bottleneck and optimize later? We could probably use SIMD types to force generate better code

Reply:
yeah I was thinking about this, calculating over single distance table may lead to better cache locality here instead of processing 2 at the same time.
for high PQ, it doesn't matter because the distance table is 2D matrix of shape (num_sub_vector, num_centroids), for the same sub vector, it's always 16 length (2^4)